### PR TITLE
[LangOpts] NFC: Future proof language version checking

### PR DESCRIPTION
**Pull request by davezarzycki**

### CODE REVIEW

**Comment by davezarzycki** [comment]
Let's try and see how swift CI works again:

@swift-ci please smoke test

File: lib/ClangImporter/ClangAdapter.cpp [hunk 1]
```diff
@@ -712,7 +712,7 @@ OptionalTypeKind importer::getParamOptionality(version::Version swiftVersion,
     return OTK_None;
 
   // Check for the 'static' annotation on C arrays.
-  if (!swiftVersion.isVersion3())
+  if (swiftVersion != version::Version({3}))
```
Comment (DougGregor) [inline lib/ClangImporter/ClangAdapter.cpp]: Wouldn't this be better as `swiftVersion >= 4`?

File: include/swift/AST/ASTContext.h [hunk 2]
```diff
@@ -863,8 +863,13 @@ class ASTContext {
...
   CanGenericSignature getSingleGenericParameterSignature() const;
   
+  /// Whether our effective Swift version is less than the given version
+  bool isSwiftVersionLessThan(unsigned major, unsigned minor = 0) const {
+    return LangOpts.isSwiftVersionLessThan(major, minor);
+  }
+
   /// Whether our effective Swift version is in the Swift 3 family
-  bool isSwiftVersion3() const { return LangOpts.isSwiftVersion3(); }
+  bool isSwiftVersion3() const { return isSwiftVersionLessThan(4); }
```
Comment (DougGregor) [inline include/swift/AST/ASTContext.h]: Definitely an improvement, but I still don't like the name `isSwiftVersion3()`; `isSwiftVersionLessThan(4)` or `< 4` would be far clearer, IMO.

Comment (davezarzycki) [inline include/swift/AST/ASTContext.h]: My original patch removed isSwiftVersion3() entirely, but Ewa wanted something more conservative. I'm open to "< 4" too. What would you prefer?

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

**Comment by gottesmm** [comment]
@davezarzycki Just an FYI, when you do a force push there is a bug (IIRC) in the jenkins plugin that we are using that means you just have to give it the command twice. @shahmishal is my memory right?

**Comment by davezarzycki** [comment]
Thanks. As it turns out, I can reproduce this failure locally. But I'm glad to learn about the CI bug.

File: lib/ClangImporter/ClangAdapter.cpp (hunk 1 above)
Comment (ematejska) [inline lib/ClangImporter/ClangAdapter.cpp]: I would prefer to leave this exactly as is, unperturbed and add the new utility function for future use.  What do you all think?  This means, this is reverted to what it was.

File: include/swift/Basic/LangOptions.h [hunk 3]
```diff
@@ -306,11 +306,14 @@ namespace swift {
...
-    /// Whether our effective Swift version is in the Swift 3 family
-    bool isSwiftVersion3() const {
-      return EffectiveLanguageVersion.isVersion3();
+    /// Whether our effective Swift version is less than "X.Y"
+    bool isSwiftVersionLessThan(unsigned major, unsigned minor = 0) const {
+      return version::Version({major, minor}) >= EffectiveLanguageVersion;
     }
 
+    /// Whether our effective Swift version is in the Swift 3 family
+    bool isSwiftVersion3() const { return isSwiftVersionLessThan(4); }
```
Comment (ematejska) [inline include/swift/Basic/LangOptions.h]: I would prefer to leave this exactly as is, unperturbed.  This means, revert isSwiftVersion3 to what it was and only add new code for the isSwiftversionLessThen function.

**Comment by gottesmm** [comment]
Hmmm... maybe we should add a note about the bug to ContinuousIntegration.md. I will file an SR.

File: include/swift/Basic/Version.h [hunk 4]
```diff
@@ -107,9 +107,6 @@ class Version {
   /// compiler to act as if it is version 3.1.
   Optional<Version> getEffectiveLanguageVersion() const;
 
-  /// Whether this version is in the Swift 3 family
-  bool isVersion3() const { return !empty() && Components[0] == 3; }
```
Comment (ematejska) [inline include/swift/Basic/Version.h]: I would prefer to leave this unperturbed.  Which means, don't remove this.

File: lib/ClangImporter/ClangAdapter.cpp (hunk 1 above)
Comment (ematejska) [inline lib/ClangImporter/ClangAdapter.cpp]: I would prefer to leave this unperturbed, which means back out this part of the change.  The idea is to just add the utility function isSwiftVersionLessThan for future code use and not perturb the existing code using isSwiftVersion3.

File: include/swift/AST/ASTContext.h (hunk 2 above)
Comment (gspiers) [reply include/swift/AST/ASTContext.h]: @davezarzycki I was wondering if you were still working on adding this to Swift in another pull request? I need to conditionally act on the Swift mode (4 or 5) for another fix I'm working on. Thanks

Comment (davezarzycki) [reply include/swift/AST/ASTContext.h]: Hi @gspiers. I'm am not. This should have been a very trivial change, but after spending too many hours trying to debug a regression from this change in the Swift 3 to 4 migrator, I've handed off the work to @ematejska.

Comment (gspiers) [reply include/swift/AST/ASTContext.h]: Thanks @davezarzycki for the quick reply. I'll keep an eye out for the changes to land in master before I submit my fix.

**Pull request closed without merging**